lib/std/heap/binned_allocator.zig

```diff
@@ -0,0 +1,818 @@
+//! # Binned Allocator
+//!
+//! This is designed to be a fast, general-purpose allocator, for use in place of GeneralPurposeAllocator when speed is the priority.
+//! It uses a collection of "bins" for various sizes of allocation, combined with a binary tree for allocations larger than 4kiB.
+//!
+//! Features:
+//!  - zero per-allocation metadata
+//!  - optional thread-safety (on by default; disable for improved single-threaded performance)
+//!  - primitive leak checking (for full checking, use GeneralPurposeAllocator)
+//!  - does not free to backing allocator, allowing use with FixedBufferAllocator
+//!
+//! Limitations:
+//!  - allocations will be at least as big as their alignment
+//!  - allocation sizes will be rounded up to the nearest bin size
+//!  - alignment cannot be greater than std.mem.page_size (see also #7952)
+//!  - does not free to backing allocator, leading to memory leaks for some niche allocation patterns
```

Comment:
> leading to memory leaks for some niche allocation patterns

This deserves to be explained, ideally in what cases it is not usable, but at least in all positive cases. Or should this be deleted?

Reply:
Yeah this comment is probably not very helpful. The main gotcha is fragmentation, since any memory locked up between small allocations will never be freed.
The other potential problem case is if you allocate and deallocate objects of gradually increasing sizes, the smaller ones will never be freed, but that problem disappears as soon as the allocations get larger than the page size so it's not worth worrying about in most cases.

As the comment suggests, they're not common cases so not generally worth worrying about so I'm not sure I should go into too much depth, but it's probably worth at least keeping a warning there for people with significant memory constraints. To be honest the whole allocator isn't all that memory efficient so maybe I should just warn about memory usage in general instead.